dhctl/pkg/operations/bootstrap/post-script.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2022 Flant JSC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package bootstrap
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"time"
+
+	"github.com/deckhouse/deckhouse/dhctl/pkg/log"
+	"github.com/deckhouse/deckhouse/dhctl/pkg/system/ssh"
+)
+
+type PostBootstrapScriptExecutor struct {
+	path        string
+	timeout     time.Duration
+	errorIfFail bool
+	sshClient   *ssh.Client
+	state       *State
+}
+
+func NewPostBootstrapScriptExecutor(sshClient *ssh.Client, path string, state *State) *PostBootstrapScriptExecutor {
+	return &PostBootstrapScriptExecutor{
+		path:      path,
+		sshClient: sshClient,
+		state:     state,
+	}
+}
+
+func (e *PostBootstrapScriptExecutor) WithTimeout(timeout time.Duration) *PostBootstrapScriptExecutor {
+	e.timeout = timeout
+	return e
+}
+
+func (e *PostBootstrapScriptExecutor) WithErrorIfExecutionFail(f bool) *PostBootstrapScriptExecutor {
+	e.errorIfFail = f
+	return e
+}
+
+func (e *PostBootstrapScriptExecutor) Execute() error {
+	return log.Process("bootstrap", "Execute post-bootstrap script", func() error {
+		var err error
+		resultToSetState, err := e.run()
+
+		if err != nil {
+			msg := fmt.Sprintf("Post execution script was failed: %v", err)
+			if e.errorIfFail {
+				return errors.New(msg)
+			}
+
+			log.ErrorF(msg)
+
+			return nil
+		}
+
+		err = e.state.SavePostBootstrapScriptResult(resultToSetState)
+		if err != nil {
+			log.ErrorF("Post bootstrap script result was not saved: %v", err)
+		}
+
+		return nil
+	})
+}
+
+var resultPattern = regexp.MustCompile("(?m)^Result of post-bootstrap script:(.+)$")
+
+func (e *PostBootstrapScriptExecutor) run() (string, error) {
+	var result string
+	cmd := e.sshClient.UploadScript(e.path).
+		WithSetExecuteModeBefore(true).
+		WithTimeout(e.timeout).
+		Sudo()
+
+	out, err := cmd.Execute()
```

Comment:
I assume that live output should be the default for post bootstrap scripts, or one of the options at least.

Reply:
1. We need output for logging.
2. We need output for save results. See [this comment](https://github.com/deckhouse/deckhouse/pull/890#discussion_r816927089)

And we need it for prevent `merged` output from another jobs.

Reply:
1. This is ok. The live output presumes that all entries will be in a log.
2. Yes, that's why we need to abstain from caching and make the output live. Users should see what's going on in the cluster. I assume that the postscript will be used to discover some data from the cluster, and if I were a user, I would like to see discovery attempts results.

Reply:
Nikolay will debug the problem a little more. If there is no success, he will leave a to-do message, and we will merge it as is.